pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,232 @@
+	newBlock := oldBlock &^ mask
+	if oldBlock != newBlock {
+		b.bitBlock[blockNum] = newBlock
+		b.numOnes--
+	}
+}
+
+// Clone the Bitmap.
+func (b *Bitmap) Clone() *Bitmap {
+	ptr := Bitmap{b.numOnes, make([]uint64, len(b.bitBlock))}
+	copy(ptr.bitBlock, b.bitBlock[:])
+	return &ptr
+}
+
+// ReplaceWith replace the content of b with that of newBitmap.
+func (b *Bitmap) ReplaceWith(newBitmap *Bitmap) {
+	b.numOnes = newBitmap.numOnes
+	b.bitBlock = make([]uint64, len(newBitmap.bitBlock))
+	copy(b.bitBlock, newBitmap.bitBlock[:])
+}
+
+// countOnesForBlocks count all 1 bits within b.bitBlock of begin and that of end.
+// The begin block and end block are inclusive.
+func (b *Bitmap) countOnesForBlocks(begin, end int) uint64 {
+	ones := uint64(0)
+	if begin > end {
+		return ones
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	for i := beginBlock; i <= endBlock; i++ {
+		ones += uint64(bits.OnesCount64(b.bitBlock[i]))
+	}
+	return ones
+}
+
+// countOnesForAllBlocks count all 1 bits in b.bitBlock.
+func (b *Bitmap) countOnesForAllBlocks() uint64 {
+	ones := uint64(0)
+	for i := 0; i < len(b.bitBlock); i++ {
+		ones += uint64(bits.OnesCount64(b.bitBlock[i]))
+	}
+	return ones
+}
+
+// flipBitmapRange flip the bits within range (begin and end). begin and end are inclusive.
+func (b *Bitmap) flipBitmapRange(begin, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		b.bitBlock[blockNum] = b.bitBlock[blockNum] ^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		b.bitBlock[endBlock] ^= ((^uint64(0) << uint(begin%64)) & ((uint64(1) << (uint(end) % 64 + 1))-1))
+	} else {
+		b.bitBlock[beginBlock] ^= ^(^uint64(0) << uint(begin%64))
+		for i := beginBlock; i < endBlock; i++ {
+			b.bitBlock[i] = ^b.bitBlock[i]
+		}
+		b.bitBlock[endBlock] ^= ((uint64(1) << (uint(end) % 64 + 1))-1)
+	}
+}
+
+// removeBitmapRange remove the bits within range (begin and end). begin and end are inclusive.
+func (b *Bitmap) removeBitmapRange(begin, end int) {
+	if begin > end {
+		return
+	}
+	if begin == end {
+		blockNum, mask := begin/64, uint64(1)<<(begin%64)
+		b.bitBlock[blockNum] = b.bitBlock[blockNum] &^ mask
+		return
+	}
+	beginBlock := begin / 64
+	endBlock := end / 64
+	if beginBlock == endBlock {
+		b.bitBlock[beginBlock] &= (((uint64(1) << uint(begin%64)) - 1) | ^((uint64(1) << (uint(end) % 64 + 1)) - 1))
+	} else {
+		b.bitBlock[beginBlock] &= ((uint64(1) << uint(begin%64)) - 1)
+		for i := beginBlock + 1; i < endBlock; i++ {
+			b.bitBlock[i] &= ^b.bitBlock[i]
+		}
+		b.bitBlock[endBlock] &= ^((uint64(1) << (uint(end) % 64 + 1)) - 1)
+	}
+}
+
+// RemoveRange remove bits within range (begin and end) for the Bitmap. begin and end are inclusive.
+func (b *Bitmap) RemoveRange(begin, end int) {
+	blockRange := end/64 - begin/64
+	// when the number of removed blocks is larger than half of the length of b.bitBlock
```

Comment:
nit: All comments should be complete sentences and grammatically correct. We try to hold this standard through out our codebase. So here capitalize the first letter "When" and add a period at the end. Same for all the comments added in this change.

Reply:
> add a period at the end

I am not sure about this comment.

Reply:
All sentences in comments should end with a full stop (period or .). For example, a function comment was "compSlice return true if slice a equal to slice b". It should end with a full stop "... to slice b.".

Reply:
Sorry about my poor English, I thought the "period" means "a period of time".